# [E-02} 과제

## 손글씨 분류하기

### 데이터
#### 데이터 준비

In [2]:
# 필요한 모듈 Import
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

digits = load_digits()  # 인스턴스 = 클래스()
digits_data = digits.data
digits_label = digits.target

#### train, Test 데이터 분리

In [3]:
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=13)

### Model

#### Decision Tree Model  

In [5]:
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=13)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))


# 정확도(Accuracy) 확인
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       0.97      0.93      0.95        30
           1       0.83      0.78      0.81        37
           2       0.87      0.87      0.87        38
           3       0.79      0.84      0.82        37
           4       0.89      0.87      0.88        38
           5       0.89      0.93      0.91        45
           6       0.95      0.93      0.94        42
           7       0.96      0.92      0.94        26
           8       0.93      0.79      0.86        34
           9       0.72      0.88      0.79        33

    accuracy                           0.88       360
   macro avg       0.88      0.87      0.88       360
weighted avg       0.88      0.88      0.88       360



NameError: name 'confusion_matrix' is not defined

#### Random Forest 사용

In [6]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))


# 정확도(Accuracy) 확인
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.90      1.00      0.95        37
           2       0.97      1.00      0.99        38
           3       1.00      0.95      0.97        37
           4       0.97      1.00      0.99        38
           5       1.00      0.98      0.99        45
           6       1.00      0.98      0.99        42
           7       0.96      0.96      0.96        26
           8       0.93      0.79      0.86        34
           9       0.89      0.97      0.93        33

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



NameError: name 'confusion_matrix' is not defined

#### SVM Model

In [7]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))


# 정확도(Accuracy) 확인
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.93      1.00      0.96        37
           2       0.97      1.00      0.99        38
           3       1.00      0.97      0.99        37
           4       1.00      1.00      1.00        38
           5       1.00      0.98      0.99        45
           6       1.00      1.00      1.00        42
           7       1.00      0.96      0.98        26
           8       1.00      0.91      0.95        34
           9       0.94      1.00      0.97        33

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



NameError: name 'confusion_matrix' is not defined

#### SGD Model 사용하기

In [8]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))


# 정확도(Accuracy) 확인
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.79      1.00      0.88        37
           2       1.00      0.97      0.99        38
           3       0.95      0.95      0.95        37
           4       0.97      1.00      0.99        38
           5       1.00      0.89      0.94        45
           6       1.00      1.00      1.00        42
           7       0.93      0.96      0.94        26
           8       0.96      0.68      0.79        34
           9       0.89      1.00      0.94        33

    accuracy                           0.94       360
   macro avg       0.95      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360



NameError: name 'confusion_matrix' is not defined

#### Logistic Regression Model

In [9]:

from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print("######정답과 오답을 구분해서 확인#######")
print(confusion_matrix(y_test, y_pred))


# 정확도(Accuracy) 확인
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.85      0.92      0.88        37
           2       1.00      0.95      0.97        38
           3       0.97      0.92      0.94        37
           4       0.95      1.00      0.97        38
           5       0.98      0.96      0.97        45
           6       1.00      0.98      0.99        42
           7       0.96      0.92      0.94        26
           8       0.97      0.82      0.89        34
           9       0.82      1.00      0.90        33

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360

######정답과 오답을 구분해서 확인#######


/Users/baekgun/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


NameError: name 'confusion_matrix' is not defined

### 모델 선정 

#### 정리
Recall : 실제로 True인 데이터를 모델이 True 라고 인식한 데이터수
Precision : 모델이 True로 예측한 데이터 중 실제로 True인 데이터수
F1 Score : Recall 값과 Precision의 값과 차이가 적을 수록 노이즈가 적은 것.

Decision Tree Model : 
 - Precision : 0.88
 - Recall    : 0.88
 - F1 score  : 0.88
 - Accuracy  :0.875
 - 우선 고려 항목 : 
 
Random forest Model : 의견 통합용
 - Precision : 0.97
 - Recall    : 0.96
 - F1 score  : 0.96
 - Accuracy  :0.963
 - 우선 고려 항목 : 
 
SVM Model
 - Precision : 0.98
 - Recall    : 0.98
 - F1 score  : 0.98
 - Accuracy  :0.983
 - 우선 고려 항목 : 
 
SGD Model
 - Precision : 0.95
 - Recall    : 0.95
 - F1 score  : 0.95
 - Accuracy  :0.947
 - 우선 고려 항목 : 
 
Logistic Regreession Model
 - Precision : 0.95
 - Recall    : 0.95
 - F1 score  : 0.95
 - Accuracy  :0.947
 - 우선 고려 항목 : 
    
### 최종 선정 모델


##### 와인 분류와 유방암 진단 분류는 동일 주소 깃허브에 같이 업로드되어 있습니다.